File: src/services/validationEngine.js
/**
 * validationEngine.js
 * ───────────────────
 * Orchestrates the full hallucination-detection pipeline:
 *
 *   Input → Claim Extraction → Evidence Validation → Contradiction Detection
 *         → Confidence Scoring → Hallucination Classification → Final Report
 *
 * This module is the single entry point consumed by the UI.
 */

import { extractClaims }        from './claimExtractor.js';
import { validateEvidence }     from './evidenceValidator.js';
import { detectContradiction }  from './hallucinationDetector.js';
import { scoreConfidence, classifyStatus, generateExplanation } from './confidenceScorer.js';

/**
 * Process a single claim through the full pipeline.
 *
 * @param {{ id: number, text: string, type: string }} claim
 * @param {string} referenceText
 * @returns {object} enriched claim result
 */
function processClaim(claim, referenceText) {
  const hasReference = Boolean(referenceText && referenceText.trim());

  // Step 1 — Evidence retrieval & semantic similarity
  const { similarityScore, evidence, matched } =
    validateEvidence(claim.text, referenceText);

  // Step 2 — Contradiction detection
  const { contradictionScore } =
    detectContradiction(claim.text, referenceText, similarityScore);

  // Step 3 — Confidence scoring
  const { confidence, pct } =
    scoreConfidence({ similarityScore, contradictionScore, hasReference });

  // Step 4 — Status classification
  const status = classifyStatus(similarityScore, contradictionScore, hasReference);

  // Step 5 — Human-readable explanation
  const explanation = generateExplanation(status, pct, hasReference);

  return {
    ...claim,
    status,
    confidence,
    confidencePct: pct,
    similarityScore: Math.round(similarityScore * 100) / 100,
    contradictionScore: Math.round(contradictionScore * 100) / 100,
    evidence,
    matched,
    explanation,
  };
}

/**
 * Compute the overall reliability score for the full response (0–100).
 *
 * Weights:
 *   Supported     → contributes full confidence value
 *   Uncertain     → contributes 50% of confidence value
 *   Unsupported   → contributes 25% of confidence value
 *   Contradictory → contributes 0 (penalizes the score)
 *
 * @param {object[]} results - Array of processed claims
 * @returns {number} 0–100
 */
function computeOverallScore(results) {
  if (results.length === 0) return 0;

  let weightedSum = 0;
  for (const r of results) {
    switch (r.status) {
      case 'supported':     weightedSum += r.confidencePct * 1.00; break;
      case 'uncertain':     weightedSum += r.confidencePct * 0.50; break;
      case 'unsupported':   weightedSum += r.confidencePct * 0.25; break;
      case 'contradictory': weightedSum += 0;                      break;
    }
  }

  return Math.round(weightedSum / results.length);
}

/**
 * Generate a human-readable validation summary paragraph.
 * @param {object[]} results
 * @param {number} score
 * @returns {string}
 */
function generateSummary(results, score) {
  const total        = results.length;
  const supported    = results.filter(r => r.status === 'supported').length;
  const unsupported  = results.filter(r => r.status === 'unsupported').length;
  const contradictory= results.filter(r => r.status === 'contradictory').length;
  const uncertain    = results.filter(r => r.status === 'uncertain').length;

  let reliability = 'needs verification';
  if (score >= 90)      reliability = 'highly reliable';
  else if (score >= 75) reliability = 'mostly reliable';
  else if (score < 50)  reliability = 'high hallucination risk';

  const parts = [`The AI response contains ${total} factual claim${total !== 1 ? 's' : ''}.`];

  if (supported > 0)
    parts.push(`${supported} claim${supported !== 1 ? 's are' : ' is'} supported by the available evidence.`);
  if (unsupported > 0)
    parts.push(`${unsupported} claim${unsupported !== 1 ? 's lack' : ' lacks'} sufficient evidence for verification.`);
  if (contradictory > 0)
    parts.push(`${contradictory} claim${contradictory !== 1 ? 's appear' : ' appears'} to contradict the trusted reference information and may represent hallucinations.`);
  if (uncertain > 0)
    parts.push(`${uncertain} claim${uncertain !== 1 ? 's are' : ' is'} flagged as uncertain due to ambiguous evidence.`);

  parts.push(`The overall reliability score is ${score}/100 (${reliability}).`);

  if (contradictory > 0 || unsupported > 0) {
    parts.push('Manual verification is recommended before relying on this response.');
  }

  return parts.join(' ');
}

/**
 * Main export: run the complete validation pipeline.
 *
 * @param {object} input
 * @param {string} input.query          - User's original question
 * @param {string} input.aiResponse     - AI-generated response to validate
 * @param {string} [input.reference=''] - Optional trusted reference context
 * @returns {{
 *   claims: object[],
 *   overallScore: number,
 *   summary: string,
 *   stats: { supported, unsupported, contradictory, uncertain, hallucinationRisk },
 *   timestamp: string,
 * }}
 */
export function runValidation({ query, aiResponse, reference = '' }) {
  // Extract individual factual claims
  const rawClaims = extractClaims(aiResponse);

  // Process each claim through the pipeline
  const results = rawClaims.map(claim => processClaim(claim, reference));

  // Compute overall score
  const overallScore = computeOverallScore(results);

  // Build stats
  const supported    = results.filter(r => r.status === 'supported').length;
  const unsupported  = results.filter(r => r.status === 'unsupported').length;
  const contradictory= results.filter(r => r.status === 'contradictory').length;
  const uncertain    = results.filter(r => r.status === 'uncertain').length;
  const hallucinationRisk = Math.round((contradictory / Math.max(1, results.length)) * 100);

  const summary = generateSummary(results, overallScore);

  return {
    query,
    aiResponse,
    reference,
    claims: results,
    overallScore,
    summary,
    stats: { supported, unsupported, contradictory, uncertain, hallucinationRisk },
    timestamp: new Date().toISOString(),
  };
}

File: backend/services/validation_engine.py
"""
validation_engine.py
────────────────────
Master orchestrator for the hallucination detection pipeline.

Full pipeline:
  Input
    → Claim Extraction        (NLTK)
    → RAG Retrieval           (LangChain + ChromaDB)
    → Semantic Similarity     (SentenceTransformers)
    → Contradiction Detection (negation + topic mismatch)
    → Confidence Scoring      (weighted formula)
    → Hallucination Classification
    → Validation Report

This is the "AI Evaluation Agent Layer" in the architecture diagram.
"""

import uuid
import time
import logging
from datetime import datetime, timezone

from services.claim_extractor      import extract_claims
from services.rag_pipeline         import RAGPipeline
from services.hallucination_detector import detect_contradiction
from services.confidence_scorer    import (
    score_confidence,
    classify_status,
    generate_explanation,
    compute_overall_score,
    get_label,
    generate_summary,
)
from services.relevance_judge      import evaluate_relevance
from services.accuracy_judge       import evaluate_accuracy
from services.hallucination_judge  import evaluate_hallucinations

logger = logging.getLogger(__name__)


def _process_claim(claim: dict, rag: RAGPipeline, has_reference: bool) -> dict:
    """
    Run the full validation pipeline for a single claim.

    Steps:
      1. RAG retrieval → best evidence + similarity score
      2. Contradiction detection
      3. Confidence scoring
      4. Status classification
      5. Explanation generation

    Returns enriched claim dict.
    """
    claim_text = claim["text"]

    # Step 1 — RAG evidence retrieval
    similarity_score, evidence = rag.retrieve_evidence(claim_text)

    # Step 2 — Contradiction detection
    ref_text = rag.reference_text or ""
    contradiction_result = detect_contradiction(claim_text, ref_text, similarity_score)
    contradiction_score  = contradiction_result["contradiction_score"]

    # Step 3 — Confidence scoring
    conf_result = score_confidence(similarity_score, contradiction_score, has_reference)

    # Step 4 — Classification
    status = classify_status(similarity_score, contradiction_score, has_reference)

    # Step 5 — Explanation
    explanation = generate_explanation(status, conf_result["pct"], has_reference)

    return {
        **claim,
        "status":              status,
        "confidence_pct":      conf_result["pct"],
        "similarity_score":    round(similarity_score, 3),
        "contradiction_score": round(contradiction_score, 3),
        "evidence":            evidence,
        "explanation":         explanation,
    }


def run_validation(query: str, ai_response: str, reference: str = "") -> dict:
    """
    Execute the complete validation pipeline with Judge Agents.

    Args:
        query:       The original user question
        ai_response: The AI-generated response to validate
        reference:   Optional trusted reference / context text

    Returns:
        Full validation report dict compatible with ValidationResponse schema
    """
    start_time   = time.time()
    record_id    = str(uuid.uuid4())
    has_reference = bool(reference and reference.strip())

    logger.info("Starting validation [id=%s] — has_reference=%s", record_id, has_reference)

    # ── Step 1: Extract claims ────────────────────────────────────────────────
    raw_claims = extract_claims(ai_response)
    logger.info("Extracted %d claims", len(raw_claims))

    # ── Step 2: Build RAG pipeline ───────────────────────────────────────────
    rag = RAGPipeline(reference)

    # ── Step 3: Judge Agent Evaluations ──────────────────────────────────────
    relevance_eval = evaluate_relevance(query, ai_response)
    accuracy_eval = evaluate_accuracy(query, ai_response, reference, rag)
    hallucination_eval = evaluate_hallucinations(ai_response, reference, rag)

    # ── Step 4: Process each claim ───────────────────────────────────────────
    results = []
    for claim in raw_claims:
        try:
            processed = _process_claim(claim, rag, has_reference)
            results.append(processed)
        except Exception as e:
            logger.error("Error processing claim %d: %s", claim["id"], e)
            # Include claim with default values on error
            results.append({
                **claim,
                "status":              "unsupported",
                "confidence_pct":      30,
                "similarity_score":    0.0,
                "contradiction_score": 0.0,
                "evidence":            "Processing error occurred.",
                "explanation":         "An error occurred while processing this claim.",
            })

    # ── Step 5: Cleanup resources ────────────────────────────────────────────
    rag.cleanup()

    # ── Step 6: Aggregate results ────────────────────────────────────────────
    overall_score  = compute_overall_score(results)
    label          = get_label(overall_score)
    summary        = generate_summary(results, overall_score)

    supported     = sum(1 for r in results if r["status"] == "supported")
    unsupported   = sum(1 for r in results if r["status"] == "unsupported")
    contradictory = sum(1 for r in results if r["status"] == "contradictory")
    uncertain     = sum(1 for r in results if r["status"] == "uncertain")
    hallucination_risk = hallucination_eval.hallucination_score

    elapsed_ms = round((time.time() - start_time) * 1000)

    report = {
        "id":            record_id,
        "query":         query,
        "ai_response":   ai_response,
        "reference":     reference,
        "claims":        results,
        "overall_score": overall_score,
        "summary":       summary,
        "label":         label,
        "stats": {
            "supported":         supported,
            "unsupported":       unsupported,
            "contradictory":     contradictory,
            "uncertain":         uncertain,
            "hallucination_risk": hallucination_risk,
        },
        "timestamp":          datetime.now(timezone.utc).isoformat(),
        "processing_time_ms": elapsed_ms,
        "relevance_eval":     relevance_eval.dict(),
        "accuracy_eval":      accuracy_eval.dict(),
        "hallucination_eval": hallucination_eval.dict(),
    }

    logger.info(
        "Validation complete [id=%s] score=%d label=%s elapsed=%dms",
        record_id, overall_score, label, elapsed_ms
    )

    return report

